### PR TITLE
Fixed Hostmanager concurrent test

### PR DESCRIPTION
**Pull request by gcatron**
*Description*: Fixed HostManager concurrent test and added additional concurrent test. 
Added method to Provisioner to removed unused compiledFunctions.
*Testing*: ninja test passed
*Documentation*: NA

### CODE REVIEW

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 1]
```diff
@@ -115,3 +115,10 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module) {
   }
   return llvm::Error::success();
 };
+
+void Provisioner::removeFunction(llvm::StringRef name) {
+  auto it = functions_.find(name);
+  if (it != functions_.end()) {
+    functions_.erase(it);
```
Comment (rdzhabarov) [inline lib/Runtime/Provisioner/Provisioner.cpp]: no need to check existence prior `erase`.

File: tests/unittests/HostManagerTest.cpp [hunk 2]
```diff
@@ -130,27 +130,40 @@ TEST_F(HostManagerTest, runNetwork) {
...
+      for (auto j = 0; j < numItersPerThread; ++j) {
+        addAndRemoveNetwork(hostManager.get(), ++counter);
+      }
+    });
+  }
+
+  for (auto &t : threads) {
+    t.join();
+  }
+}
```
Comment (rdzhabarov) [inline tests/unittests/HostManagerTest.cpp]: what exactly is tested here? (no tsan/asan failures?)

File: lib/Runtime/HostManager/HostManager.cpp [hunk 3]
```diff
@@ -57,6 +57,15 @@ HostManager::~HostManager() { llvm::toString(clearHost()); }
 
 llvm::Error HostManager::addNetwork(Module *M) {
   std::lock_guard<std::mutex> networkLock(networkLock_);
+  auto functions = M->getFunctions();
+  for (auto &F : functions) {
+    std::string name = F->getName();
+    auto it = networks_.find(name);
+    if (it != networks_.end()) {
+      return MAKE_ERR("Failed to add network: already have a function called " +
```
Comment (jackm321) [inline lib/Runtime/HostManager/HostManager.cpp]: Can you also add the generic runtime error GlowError error code

File: lib/Runtime/HostManager/HostManager.cpp [hunk 4]
```diff
@@ -90,6 +99,18 @@ void HostManager::removeNetwork(llvm::StringRef networkName) {
   if (networkIterator == networks_.end()) {
     return;
   }
+  auto &nodes = networkIterator->second.nodes;
+  for (auto &node : nodes) {
+    std::promise<llvm::Error> removeNetwork;
```
Comment (jackm321) [inline lib/Runtime/HostManager/HostManager.cpp]: This seems to cause issues for Windows (right @ayermolo?) so let's just pass an llvm::Error in by ref (can be initialized to success).

Comment (ayermolo) [reply lib/Runtime/HostManager/HostManager.cpp]: Yep. Relevant discussion: https://github.com/pytorch/glow/pull/2477

File: tests/unittests/HostManagerTest.cpp (hunk 2 above)
Comment (gcatron) [reply tests/unittests/HostManagerTest.cpp]: Yes, no TSAN/ASAN failures, and general stability.

Comment (gcatron) [reply tests/unittests/HostManagerTest.cpp]: I'll add comments above the tests.

File: lib/Runtime/HostManager/HostManager.cpp (hunk 4 above)
Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: What part causes issues? Using an error in a promise?

Comment (ayermolo) [reply lib/Runtime/HostManager/HostManager.cpp]: Two issues.
1) using llvm error in promise results in compilation error in visual studio 2017 with llvm 7.0.0
2) There is run time stack corruption when llvm gets executed.

The conclusion seems to be templated std::promise windows implementation has issues. @nickgg  proposed a solution that seems to work.

Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: Thanks, I'll use promise<void>.

File: include/glow/Runtime/Provisioner/Provisioner.h [hunk 5]
```diff
@@ -39,6 +39,9 @@ class Provisioner final {
   /// success.
   llvm::Error provision(DAGListTy &networks, Module &module);
 
+  /// Remove stored compiledFunction
```
Comment (rdzhabarov) [inline include/glow/Runtime/Provisioner/Provisioner.h]: '.' :D

File: lib/Runtime/HostManager/HostManager.cpp (hunk 4 above)
Comment (nickgg) [reply lib/Runtime/HostManager/HostManager.cpp]: nit: you can use backticks and do `std::promise<void>` 😃

Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: I think it should be good to go now.

File: lib/Runtime/HostManager/HostManager.cpp [hunk 6]
```diff
@@ -57,6 +57,16 @@ HostManager::~HostManager() { llvm::toString(clearHost()); }
 
 llvm::Error HostManager::addNetwork(Module *M) {
   std::lock_guard<std::mutex> networkLock(networkLock_);
+  auto functions = M->getFunctions();
+  for (auto &F : functions) {
+    std::string name = F->getName();
+    auto it = networks_.find(name);
+    if (it != networks_.end()) {
+      return MAKE_ERR(GlowErr::ErrorCode::RUNTIME_REQUEST_REFUSED,
```
Comment (jackm321) [inline lib/Runtime/HostManager/HostManager.cpp]: `RUNTIME_REQUEST_REFUSED` is for inference requests this is for adding network, just `RUNTIME_ERROR` would be better

**Pull request merged**